language_service/src/tests.rs

```diff
@@ -636,13 +636,50 @@ fn new_language_service(received: &RefCell<Vec<ErrorInfo>>) -> LanguageService<'
     LanguageService::new(
         diagnostic_receiver,
         // these tests do not test project mode
-        // so we provide
+        // so we provide these stubbed-out functions
         |_| Box::pin(std::future::ready((Arc::from(""), Arc::from("")))),
         |_| Box::pin(std::future::ready(vec![])),
         |_| Box::pin(std::future::ready(None)),
     )
 }
 
+// the below tests test the asynchronous behavior of the language service.
+// we use `get_completions` as a rough analog for all document operations, as
+// they all go through the same `document_op` infrastructure.
+#[tokio::test]
+async fn completions_requested_before_document_load() {
+    let errors = RefCell::new(Vec::new());
+    let mut ls = new_language_service(&errors);
+
+    // we intentionally don't await this to test how LSP features function when
+    // a document hasn't fully loaded
+    let _ = ls.update_document(
+        "foo.qs",
+        1,
+        "namespace Foo { open Microsoft.Quantum.Diagnostics; @EntryPoint() operation Main() : Unit { DumpMachine() } }",
+    );
+
+    // this should be empty, because the doc hasn't loaded
+    assert!(ls.get_completions("foo.qs", 76).items.is_empty())
+}
```

Comment:
You may want to hold on to the future from `update_document` and `await` on it here, just to make sure the _test_ waits for its completion.